File: Sprint6/F/C++/main.cpp
// https://contest.yandex.ru/contest/25069/problems/F/
#include <iostream>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

vector<vector<int>> graph;

int min_dist_between_two_vertices(int start, int end) {
    vector<int> dist(graph.size(), INT_MAX);
    dist[start] = 0;

    queue<int> q;
    q.push(start);

    while (!q.empty()) {
        int current = q.front();
        q.pop();

        for (int neighbor : graph[current]) {
            if (dist[neighbor] == INT_MAX) {
                dist[neighbor] = dist[current] + 1;
                q.push(neighbor);
            }
        }
    }
    return dist[end] == INT_MAX ? -1 : dist[end];
}


int main() {
    int n, m;
    cin >> n >> m;

    graph.resize(n);

    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        --u;
        --v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    int start, end;
    cin >> start >> end;

    cout << min_dist_between_two_vertices(--start, --end) << endl;

    return 0;
}

File: Sprint6/F/Rust/src/main.rs
// https://contest.yandex.ru/contest/25069/problems/F/

use std::collections::VecDeque;
use std::io::{self, BufRead};

fn min_dist_between_two_vertices(graph: &Vec<Vec<usize>>, start: usize, end: usize) -> i32 {
    let mut dist = vec![i32::MAX; graph.len()];
    dist[start] = 0;

    let mut q: VecDeque<usize> = VecDeque::new();
    q.push_back(start);

    while let Some(current) = q.pop_front() {
        for &neighbor in &graph[current] {
            if dist[neighbor] == i32::MAX {
                dist[neighbor] = dist[current] + 1;
                q.push_back(neighbor);
            }
        }
    }
    if dist[end] == i32::MAX {
        -1
    } else {
        dist[end]
    }
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let first_line = lines.next().unwrap().unwrap();
    let nm: Vec<usize> = first_line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    let (n, m) = (nm[0], nm[1]);

    let mut graph = vec![vec![]; n];

    for _ in 0..m {
        let edge: Vec<usize> = lines
            .next()
            .unwrap()
            .unwrap()
            .split_whitespace()
            .map(|x| x.parse::<usize>().unwrap() - 1)
            .collect();

        let (u, v) = (edge[0], edge[1]);
        graph[u].push(v);
        graph[v].push(u);
    }

    let se: Vec<usize> = lines
        .next()
        .unwrap()
        .unwrap()
        .split_whitespace()
        .map(|x| x.parse::<usize>().unwrap() - 1)
        .collect();
    let (start, end) = (se[0], se[1]);

    println!("{}", min_dist_between_two_vertices(&graph, start, end));
}
